src/adam_core/propagator/propagator.py

```diff
@@ -89,17 +93,158 @@ class EphemerisMixin:
-
-        THIS FUNCTION SHOULD BE DEFINED BY THE USER.
+        A generic ephemeris implementation, which can be used or overridden by subclasses.
         """
-        pass
+
+        if isinstance(orbits, Orbits):
+            ephemeris_total = Ephemeris.empty()
+        elif isinstance(orbits, VariantOrbits):
+            ephemeris_total = VariantEphemeris.empty()
+
+        for orbit in orbits:
+            propagated_orbits = self.propagate_orbits(orbit, observers.coordinates.time)
+
+            # Transform both the orbits and observers to the barycenter if they are not already.
+            propagated_orbits_barycentric = propagated_orbits.set_column(
+                "coordinates",
+                transform_coordinates(
+                    propagated_orbits.coordinates,
+                    CartesianCoordinates,
+                    frame_out="ecliptic",
+                    origin_out=OriginCodes.SOLAR_SYSTEM_BARYCENTER,
+                ),
+            )
+            observers_barycentric = observers.set_column(
+                "coordinates",
+                transform_coordinates(
+                    observers.coordinates,
+                    CartesianCoordinates,
+                    frame_out="ecliptic",
+                    origin_out=OriginCodes.SOLAR_SYSTEM_BARYCENTER,
+                ),
+            )
+            num_orbits = len(propagated_orbits_barycentric.orbit_id.unique())
+
+            observer_codes = np.tile(
+                observers.code.to_numpy(zero_copy_only=False), num_orbits
+            )
+
+            propagated_orbits_aberrated, light_time = self._add_light_time(
+                propagated_orbits_barycentric,
+                observers_barycentric,
+                lt_tol=lt_tol,
+            )
+
+            topocentric_coordinates = CartesianCoordinates.from_kwargs(
+                x=propagated_orbits_aberrated.coordinates.values[:, 0]
+                - observers_barycentric.coordinates.values[:, 0],
+                y=propagated_orbits_aberrated.coordinates.values[:, 1]
+                - observers_barycentric.coordinates.values[:, 1],
+                z=propagated_orbits_aberrated.coordinates.values[:, 2]
+                - observers_barycentric.coordinates.values[:, 2],
+                vx=propagated_orbits_aberrated.coordinates.values[:, 3]
+                - observers_barycentric.coordinates.values[:, 3],
+                vy=propagated_orbits_aberrated.coordinates.values[:, 4]
+                - observers_barycentric.coordinates.values[:, 4],
+                vz=propagated_orbits_aberrated.coordinates.values[:, 5]
+                - observers_barycentric.coordinates.values[:, 5],
+                covariance=None,
+                # The ephemeris times are at the point of the observer,
+                # not the aberated orbit
+                time=observers.coordinates.time,
+                origin=Origin.from_kwargs(code=observer_codes),
+                frame="ecliptic",
+            )
```

Comment:
```suggestion
            topocentric_state = propagated_orbits_aberrated.coordinates.values - observers_barycentric.coordinates.values
            topocentric_coordinates = CartesianCoordinates.from_kwargs(
                x=topocentric_state[:, 0],
                y=topocentric_state[:, 1],
                z=topocentric_state[:, 2],
                vx=topocentric_state[:, 3],
                vy=topocentric_state[:, 4],
                vz=topocentric_state[:, 5],
                covariance=None,
                # The ephemeris times are at the point of the observer,
                # not the aberrated orbit
                time=observers.coordinates.time,
                origin=Origin.from_kwargs(code=observer_codes),
                frame="ecliptic",
            )
```

Probably will only get a small performance gain but this way we only call `np.array(self.table.select(["x", "y", "z", "vx", "vy", "vz"]))` once.

Reply:
Maybe one day we overload the math operators for our coordinates classes.

Reply:
> Probably will only get a small performance gain but this way we only call np.array(self.table.select(["x", "y", "z", "vx", "vy", "vz"])) once.
Twice not once but at least not 12 times!